oauth2/oauth2_refresh_token_test.go

```diff
@@ -0,0 +1,155 @@
+	// first read.
+	workers := 10
+	postgresRegistry := internal.NewRegistrySQL(internal.NewConfigurationWithDefaults(), connectToPG(t))
+	x.CleanSQL(t, postgresRegistry.DB())
+	_, err := postgresRegistry.CreateSchemas("postgres")
+	require.NoError(t, err)
+
+	token := "234c678fed33c1d2025537ae464a1ebf7d23fc4a"
+	tokenSignature := "4c7c7e8b3a77ad0c3ec846a21653c48b45dbfa31"
+	testClient := hc.Client{
+		ClientID:      uuid.New(),
+		Secret:        "secret",
+		ResponseTypes: []string{"id_token", "code", "token"},
+		GrantTypes:    []string{"implicit", "refresh_token", "authorization_code", "password", "client_credentials"},
+		Scope:         "hydra offline openid",
+		Audience:      []string{"https://api.ory.sh/"},
+	}
+
+	request := &fosite.AccessRequest{
+		GrantTypes: []string{
+			"refresh_token",
+		},
+		Request: fosite.Request{
+			ID: uuid.New(),
+			Client: &hc.Client{
+				ClientID: testClient.ClientID,
+			},
+			RequestedScope: []string{"offline"},
+			GrantedScope:   []string{"offline"},
+			Session:        oauth2.NewSession(""),
+			Form: url.Values{
+				"refresh_token": []string{fmt.Sprintf("%s.%s", token, tokenSignature)},
+			},
+		},
+	}
+
+	ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(10*time.Second))
+	require.NoError(t, postgresRegistry.OAuth2Storage().(clientCreator).CreateClient(ctx, &testClient))
+	assert.NoError(t, postgresRegistry.OAuth2Storage().CreateRefreshTokenSession(ctx, tokenSignature, request))
+	_, err = postgresRegistry.OAuth2Storage().GetRefreshTokenSession(ctx, tokenSignature, nil)
+	assert.NoError(t, err)
+	provider := postgresRegistry.OAuth2Provider()
+
+	var wg sync.WaitGroup
+	for run := 0; run < testRuns; run++ {
+		barrier := make(chan struct{})
+		errorsCh := make(chan error, workers)
+
+		go func() {
+			for w := 0; w < workers; w++ {
+				wg.Add(1)
+				go func(run, worker int) {
+					defer wg.Done()
+					ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+					// all workers will block here until the for loop above has launched all the worker go-routines
+					// this is to ensure we fire all the workers off at the same
+					<-barrier
+					_, err := provider.NewAccessResponse(ctx, request)
+					errorsCh <- err
+				}(run, w)
+			}
+
+			// wait until all workers have completed their work
+			wg.Wait()
+			close(errorsCh)
+		}()
+
+		// let the race begin!
+		// all worker go-routines will now attempt to hit the "NewAccessResponse" method
+		close(barrier)
+
+		// process worker results
+
+		// successCount is the number of workers that were able to call "NewAccessResponse" without receiving an error.
+		// if the successCount at the end of a test run is bigger than one, it means that multiple access/refresh tokens
+		// were issued using the same refresh token! - https://knowyourmeme.com/memes/scared-hamster
+		var successCount int
+		for err := range errorsCh {
+			if err != nil {
+				switch err := errors.Cause(err).(type) {
```

Comment:
There's actually a bug in `errors.Cause` which can cause a panic if `Cause()` returns `nil`. This hasn't been fixed in upstream but you can use `github.com/ory/x/errorx.Cause()` as a workaround. Maybe this also has been fixed already, not sure.

Reply:
Done ✅